shared/cluster.go

```diff
@@ -409,3 +409,139 @@ func WriteDataPod(namespace string) (string, error) {
+
+const (
+	Server ServiceType = "server"
+	Agent  ServiceType = "agent"
+)
+
+func (s ServiceType) IsValidServiceType() bool {
+	valid := []ServiceType{Server, Agent}
+	for _, i := range valid {
+		if s == i {
+			return true
+		}
+	}
+
+	return false
+}
+
+type ServiceAction string
+
+const (
+	Stop    ServiceAction = "stop"
+	Start   ServiceAction = "start"
+	Restart ServiceAction = "restart"
+	Status  ServiceAction = "status"
+)
+
+func (s ServiceAction) IsValidServiceAction() bool {
+	valid := []ServiceAction{Stop, Start, Restart, Status}
+	for _, i := range valid {
+		if s == i {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (s ServiceAction) isStartAction() bool {
+	startAction := []ServiceAction{Start, Restart}
+	for _, i := range startAction {
+		if s == i {
+			return true
+		}
+	}
+
+	return false
+}
+
+type Product string
+
+const (
+	K3S  Product = "k3s"
+	RKE2 Product = "rke2"
+)
+
+func (p Product) IsValidProduct() bool {
+	valid := []Product{K3S, RKE2}
+	for _, i := range valid {
+		if p == i {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (p Product) getString() string {
+	if p == K3S {
+		return "k3s"
+	}
+
+	return "rke2"
+}
+
+// getServiceName Get service name. Used to work with stop/start k3s/rke2 services
+func (product Product) getServiceName(serviceType ServiceType) string {
+	if !serviceType.IsValidServiceType() {
+		ReturnLogError("serviceType is not valid; has to be set to ONLY: server | agent")
+	}
+	var serviceName string
+	if product == K3S && serviceType == Server {
+		serviceName = product.getString() // k3s
+	} else { // k3s-agent, rke2-server, rke2-agent
+		serviceName = fmt.Sprintf("%s-%s", product, serviceType)
+	}
+
+	return serviceName
+}
+
+func (product Product) GetServiceCmd(action ServiceAction, serviceType ServiceType) string {
+	if action.isStartAction() {
+		return fmt.Sprintf("timeout 2m sudo systemctl --no-block %s %s; sleep 60", action, product.getServiceName(serviceType))
+	}
+	return fmt.Sprintf("timeout 2m sudo systemctl --no-block %s %s", action, product.getServiceName(serviceType))
+}
+
+// ManageClusterService action:stop/start/restart/status product:rke2/k3s ips:ips array for serviceType:agent/server
+func (product Product) ManageClusterService(action ServiceAction, serviceType ServiceType, ips []string) error {
+	if !product.IsValidProduct() {
+		ReturnLogError("product needs to be one of: k3s | rke2")
+	}
+	if !action.IsValidServiceAction() {
+		ReturnLogError("action needs to be one of: start | stop | restart | status")
+	}
+	if !serviceType.IsValidServiceType() {
+		ReturnLogError("serviceType needs to be one of: server | agent")
+	}
+
+	for _, ip := range ips {
+		cmd := product.GetServiceCmd(action, serviceType)
+		_, err := RunCommandOnNode(cmd, ip)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// CertRotate certificate rotate for k3s or rke2
+func (product Product) CertRotate(ips []string) (string, error) {
+	if !product.IsValidProduct() {
```

Comment:
also fix this please - 


receiver-naming: receiver name product should be consistent with previous receiver name p for Product (revive)

Reply:
fixed and moved to product.go